### DISCUSSION TITLE
False negative in variance of dict value

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#8972] [Q&A] [answered]
(Sorry for the multiple questions these past few days).

Code sample in [pyright playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgIIA2SAhgM4A0UAslQsugFDsCSKAxgKIAPAFyESFajSgBeKAHIAJkl4wA2qhgMaMEAF0oAHyhKVq7SAY8Bgw1ABGYMOV1zO7BcWBRgAVz4AKKhA0USshAEpRBFoaN18AgG8AImJyYghiFBgaJNFkgH18qnJyQtzCEB9iAF9q8PZ43n8CopKy0QJKmvqgA)

The following does not seem to complain about the invariance of dict values, while [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=a03ed7df4044bf2f331784f8cb1261d9) does.

```python
from typing import TypeAlias, Mapping

IncEx: TypeAlias = 'dict[int, str] | dict[str, IncEx | bool]'


def func(arg: IncEx): pass


func({"elements": {"__all__": True}})
func({"__all__": True})
```

We are going to switch to `Mapping` on Pydantic but just wanted to make sure pyright wasn't missing something here. Thanks in advance.

**Comment by erictraut** [comment, marked as answer]
Pyright's behavior is correct here. The expression `{"elements": {"__all__": True}})` constructs an object that is compatible with the type `dict[str, IncEx | bool]`. This type can be evaluated through bidirectional type inference (what mypy calls "context"), so no error should be generated here.

This appears to be a bug in mypy. It looks like it is failing to apply bidirectional type inference in this case and is falling back on regular type inference rules. Pyright does this as well if you separate the code into two lines.

```python
v1: IncEx = {"elements": {"__all__": True}} # OK
reveal_type(v1) # IncEx

v2 = {"elements": {"__all__": True}}
reveal_type(v2) # dict[str, dict[str, bool]]
v3: IncEx = v2 # Type error due to invariance
```

Interestingly, you can avoid the mypy bug if you simplify the type alias slightly.

```python
from typing import TypeAlias

IncEx: TypeAlias = "dict[str, IncEx | bool]"

def func(arg: IncEx):
    pass

func({"elements": {"__all__": True}})
func({"__all__": True})
```

**Reply by Viicos** [reply]
Thanks for the answer. Indeed having `dict[int, str]` in the alias is what seems to be causing the mypy issue. We'll be using `Mapping` in the meanwhile.
